File: 0045-jump-game-ii/0045-jump-game-ii.go
func jump(nums []int) int {
    n := len(nums)
    dp := make([]int, n)
    for i := 0; i < n; i++ {
        dp[i] = 1<<31 - 1 // equivalent to INT_MAX in Go
    }
    dp[n-1] = 0
    for i := n - 2; i >= 0; i-- {
        steps := min(n-1, i + nums[i])
        for j := i + 1; j <= steps; j++ {
            if dp[j] != 1<<31 - 1 {
                dp[i] = min(dp[i], dp[j] + 1)
            }
        }
    }
    return dp[0]
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

File: 0045-jump-game-ii/0045-jump-game-ii.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var jump = function(nums) {
    let n = nums.length;
    let dp = new Array(n).fill(Number.MAX_SAFE_INTEGER);
    dp[n-1] = 0;
    for(let i=n-2; i>=0; i--) {
        let steps = Math.min(n-1, i + nums[i]);
        for(let j=i+1; j<=steps; j++) {
            if(dp[j] != Number.MAX_SAFE_INTEGER) {
                dp[i] = Math.min(dp[i], dp[j] + 1);
            }
        }
    }
    return dp[0];
};
